### PR TITLE
featured log

### PR DESCRIPTION
**Pull request by Liaojinghui**
image
image

image

image

Original:

image

### CODE REVIEW

File: neo-cli/CLI/Logger.cs [hunk 1]
```diff
@@ -11,97 +11,187 @@
...
 using Neo.Plugins;
 using System;
+using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using System.Text;
 using static System.IO.Path;
 
-namespace Neo.CLI
+namespace Neo.CLI;
```
Comment (erikzhang) [inline neo-cli/CLI/Logger.cs:20]: For new files, we can use this new style. But for old files, we should not add unnecessary changes.

File: neo-cli/CLI/Logger.cs [hunk 2]
```diff
@@ -65,43 +88,112 @@ void ILogPlugin.Log(string source, LogLevel level, object message)
...
+                                loglevel = "DEBUG";
+                                if (OperatingSystem.IsLinux() || OperatingSystem.IsMacOS())
+                                    messages[0] = $"🔨{messages[0]}";
+                                break;
+                            }
+
+                        case LogLevel.Error:
+                            logcolor = ErrorColor; loglevel = "ERROR";
+                            if (OperatingSystem.IsLinux() || OperatingSystem.IsMacOS())
+                                messages[0] = $"❌{messages[0]}";
```
Comment (shargon) [inline neo-cli/CLI/Logger.cs:115]: If the software read the output, it could produce errors, it should be possible to disable it or detect if a process is reading the output

Comment (Liaojinghui) [reply neo-cli/CLI/Logger.cs:115]: sorry i dont understand, why process reading the output could produce error here? its not yet an output but a string when i change it.

Comment (shargon) [reply neo-cli/CLI/Logger.cs:115]: It will get colors and unicode icons

Comment (Liaojinghui) [reply neo-cli/CLI/Logger.cs:115]: But i just added a few more extra operations here, the original logic still remains, if there is no issue in the past, why would be a problem here? I assume the `lock` already addresses your concern?

**Comment by erikzhang** [comment]
I don't like this style. It adds a lot of unnecessary newlines. And I don't like the icons either.

**Comment by Liaojinghui** [comment]
> I don't like this style. It adds a lot of unnecessary newlines. And I don't like the icons either.

Icon can be removed easily. new lines can be avoided. Formatted log makes it easier to address what is happening.

**Comment by Liaojinghui** [comment]
Now icons are removed, long hash characters are abbreviated

File: neo-cli/CLI/Logger.cs [hunk 3]
```diff
@@ -65,43 +89,117 @@ void ILogPlugin.Log(string source, LogLevel level, object message)
...
                     switch (level)
                     {
-                        case LogLevel.Debug: DebugColor.Apply(); break;
-                        case LogLevel.Error: ErrorColor.Apply(); break;
-                        case LogLevel.Fatal: FatalColor.Apply(); break;
-                        case LogLevel.Info: InfoColor.Apply(); break;
-                        case LogLevel.Warning: WarningColor.Apply(); break;
+                        case LogLevel.Debug:
+                            {
+#if RELEASE
```
Comment (erikzhang) [inline neo-cli/CLI/Logger.cs:103]: Why this condition?

File: neo-cli/CLI/Logger.cs [hunk 4]
```diff
@@ -1,34 +1,58 @@
...
+
+        /// <summary>
+        /// Process "log on" command
+        /// </summary>
+        [ConsoleCommand("log on", Category = "Log Commands")]
+        private void OnLogOnCommand()
+        {
+            if (Settings.Default.Logger.ConsoleOutput) showLog = true;
+            else ConsoleHelper.Warning("Please enable log first.");
+        }
```
Comment (erikzhang) [inline neo-cli/CLI/Logger.cs:54]: Why do we need this method since it is already configured in `config.json`?

Comment (Liaojinghui) [reply neo-cli/CLI/Logger.cs:54]: We need to check the log, but sometimes we also need to interact with the console, then `log` breaks our command....

File: neo-cli/CLI/Logger.cs [hunk 5]
```diff
@@ -1,34 +1,58 @@
...
+        [ConsoleCommand("log off", Category = "Log Commands")]
+        private void OnLogOffCommand()
+        {
+            showLog = false;
+        }
+
+        /// <summary>
+        /// Process "log on" command
+        /// </summary>
+        [ConsoleCommand("log on", Category = "Log Commands")]
```
Comment (erikzhang) [inline neo-cli/CLI/Logger.cs:49]: I think the command name is confused. It only control the `ConsoleOutput`, not to turn on or turn off the log actually.

Comment (Liaojinghui) [reply neo-cli/CLI/Logger.cs:49]: `console log on` and `console log off`?

File: neo-cli/CLI/Logger.cs [hunk 6]
```diff
@@ -1,34 +1,58 @@
...
         public override string Name => "SystemLog";
         public override string Description => "Prints consensus log and is a built-in plugin which cannot be uninstalled";
         public override string ConfigFile => Combine(GetDirectoryName(Path), "config.json");
         public override string Path => GetType().Assembly.Location;
+        private bool _showLog = Settings.Default.Logger.ConsoleOutput;
+
+        /// <summary>
+        /// Process "console log off" command to turn off console log
+        /// </summary>
+        [ConsoleCommand("console log off", Category = "Log Commands")]
```
Comment (erikzhang) [inline neo-cli/CLI/Logger.cs:41]: You need to register the commands.

File: neo-cli/CLI/Logger.cs [hunk 7]
```diff
@@ -65,43 +89,89 @@ void ILogPlugin.Log(string source, LogLevel level, object message)
             lock (typeof(Logger))
             {
                 DateTime now = DateTime.Now;
-                var log = $"[{now.TimeOfDay:hh\\:mm\\:ss\\.fff}] {message}";
-
-                if (Settings.Default.Logger.ConsoleOutput)
+                var log = $"[{now.TimeOfDay:hh\\:mm\\:ss\\.fff}]";
+                if (Settings.Default.Logger.ConsoleOutput && _showLog)
```
Comment (erikzhang) [inline neo-cli/CLI/Logger.cs:94]:
```suggestion
                if (_showLog)
```

File: neo-cli/CLI/Logger.cs [hunk 8]
```diff
@@ -65,43 +90,89 @@ void ILogPlugin.Log(string source, LogLevel level, object message)
...
-                        case LogLevel.Warning: WarningColor.Apply(); break;
+                        case LogLevel.Debug: logColor = DebugColor; logLevel = "DEBUG"; break;
+                        case LogLevel.Error: logColor = ErrorColor; logLevel = "ERROR"; break;
+                        case LogLevel.Fatal: logColor = FatalColor; logLevel = "FATAL"; break;
+                        case LogLevel.Info: logColor = KeyColor; logLevel = "INFO"; break;
+                        case LogLevel.Warning: logColor = WarningColor; logLevel = "WARN"; break;
+                        default: logColor = InfoColor; logLevel = "INFO"; break;
                     }
+                    logColor.Apply();
+                    Console.Write(logLevel + " ");
```
Comment (erikzhang) [inline neo-cli/CLI/Logger.cs:110]: Since we've used colors (which are the most intuitive) to represent the different categories, is this line necessary?

File: neo-cli/CLI/Logger.cs (hunk 6 above)
Comment (erikzhang) [reply neo-cli/CLI/Logger.cs:41]: Wait for #865

**Comment by erikzhang** [comment]
I'd say I like `console log on` and `console log off`. But I still prefer the original whole line in the same color, which looks more intuitive.

**Comment by nicolegys** [comment]
The log message does not write to file now.

**Comment by Liaojinghui** [comment]
@nicolegys  can you test again, i could not repeat the problem, everything works well on my machine.

**Comment by nicolegys** [comment]
> @nicolegys can you test again, i could not repeat the problem, everything works well on my machine.

File: neo-cli/CLI/MainService.Logger.cs [hunk 9]
```diff
@@ -71,43 +93,89 @@ private void OnLog(string source, LogLevel level, object message)
...
-                    path = Combine(path, $"{now:yyyy-MM-dd}.log");
-                    try
-                    {
-                        File.AppendAllLines(path, new[] { $"[{level}]{log}" });
-                    }
-                    catch (IOException)
-                    {
-                        Console.WriteLine("Error writing the log file: " + path);
-                    }
+                    File.AppendAllLines(path, new[] { $"[{level}]{log}" });
```
Comment (nicolegys) [inline neo-cli/CLI/MainService.Logger.cs:148]:
```suggestion
                    File.AppendAllLines(path, new[] { $"[{level}]{log} {message}"});
```
`message` is missing here, how could everything work well on your machine:joy:

Comment (Liaojinghui) [reply neo-cli/CLI/MainService.Logger.cs]: My bad, i must have checked the wrong log file

**Review by nicolegys — APPROVED** [review]
pass

**Comment by Liaojinghui** [comment]
@erikzhang @shargon ready to merge.

File: neo-cli/CLI/MainService.Logger.cs [hunk 10]
```diff
@@ -71,43 +93,89 @@ private void OnLog(string source, LogLevel level, object message)
...
+                        case LogLevel.Warning: logColor = WarningColor; logLevel = "WARN"; break;
+                        default: logColor = InfoColor; logLevel = "INFO"; break;
                     }
+                    logColor.Apply();
+                    Console.Write(logLevel + " ");
+                    InfoColor.Apply();
 
-                    Console.WriteLine(log);
+                    Console.Write($"{log} {messages[0]}");
+                    for (var i = 0; i < 35 - messages[0].Length - logLevel.Length; i++) Console.Write(' ');
```
Comment (shargon) [inline neo-cli/CLI/MainService.Logger.cs:117]: Only one `Console.Write`?

File: neo-cli/CLI/MainService.Logger.cs [hunk 11]
```diff
@@ -71,43 +93,89 @@ private void OnLog(string source, LogLevel level, object message)
...
+                    File.AppendAllLines(path, new[] { $"[{level}]{log} {message}" });
                 }
+                catch (IOException)
+                {
+                    Console.WriteLine("Error writing the log file: " + path);
+                }
+            }
+        }
+
+        private static string[] Parse(string message)
```
Comment (shargon) [inline neo-cli/CLI/MainService.Logger.cs:157]: Add a comment about the expected format to parse?

Comment (shargon) [reply neo-cli/CLI/MainService.Logger.cs:157]: I think we never follow any standard in messages.

Comment (Liaojinghui) [reply neo-cli/CLI/MainService.Logger.cs:157]: > I think we never follow any standard in messages.

I will work on that in my next pr after this pr is merged. I have discussed with @nicolegys  and decided to make the log system more test friendly.

**Review by erikzhang — DISMISSED** [review]
I like `console log on` and `console log off`. But I still prefer the original whole line in the same color, which looks more intuitive.

File: neo-cli/CLI/MainService.Logger.cs [hunk 12]
```diff
@@ -71,43 +93,78 @@ private void OnLog(string source, LogLevel level, object message)
...
                     {
-                        case LogLevel.Debug: DebugColor.Apply(); break;
-                        case LogLevel.Error: ErrorColor.Apply(); break;
-                        case LogLevel.Fatal: FatalColor.Apply(); break;
-                        case LogLevel.Info: InfoColor.Apply(); break;
-                        case LogLevel.Warning: WarningColor.Apply(); break;
+                        case LogLevel.Debug: logColor = DebugColor; logLevel = "DEBUG"; break;
+                        case LogLevel.Error: logColor = ErrorColor; logLevel = "ERROR"; break;
+                        case LogLevel.Fatal: logColor = FatalColor; logLevel = "FATAL"; break;
+                        case LogLevel.Info: logColor = KeyColor; logLevel = "INFO"; break;
```
Comment (nicolegys) [inline neo-cli/CLI/MainService.Logger.cs:108]:
```suggestion
                        case LogLevel.Info: logColor = InfoColor; logLevel = "INFO"; break;
```
Ok for me, except the green all over the window. Could you set the info log color as white?

File: neo-cli/CLI/MainService.Logger.cs (hunk 11 above)
Comment (nicolegys) [reply neo-cli/CLI/MainService.Logger.cs:147]: > Add a comment about the expected format to parse?

Agree, we'd better add a comment. @Liaojinghui

**Review by nicolegys — DISMISSED** [review]
Test OK. @erikzhang

File: neo-cli/CLI/MainService.Logger.cs [hunk 13]
```diff
@@ -23,8 +25,10 @@ partial class MainService
         private static readonly ConsoleColorSet WarningColor = new ConsoleColorSet(ConsoleColor.Yellow);
         private static readonly ConsoleColorSet ErrorColor = new ConsoleColorSet(ConsoleColor.Red);
         private static readonly ConsoleColorSet FatalColor = new ConsoleColorSet(ConsoleColor.Red);
+        private static readonly ConsoleColorSet KeyColor = new ConsoleColorSet(ConsoleColor.DarkGreen);
```
Comment (erikzhang) [inline neo-cli/CLI/MainService.Logger.cs:28]: This field is never used.

File: neo-cli/CLI/MainService.Logger.cs (hunk 11 above)
Comment (erikzhang) [reply neo-cli/CLI/MainService.Logger.cs:152]: Now that it's all the same color, do we still need to parse the message?

**Comment by Liaojinghui** [comment]
@shargon can you please review this pr?

File: neo-cli/CLI/MainService.Logger.cs [hunk 14]
```diff
@@ -1,30 +1,33 @@
 // Copyright (C) 2016-2022 The Neo Project.
-// 
-// The neo-cli is free software distributed under the MIT software 
+//
+// The neo-cli is free software distributed under the MIT software
 // license, see the accompanying file LICENSE in the main directory of
-// the project or http://www.opensource.org/licenses/mit-license.php
```
Comment (vncoelho) [inline neo-cli/CLI/MainService.Logger.cs:5]: These lines were format correctly? Is it needed?

**Review by vncoelho — COMMENTED** [review]
It looks much better, specially for consensus logs. I am in favor of this feature.

File: neo-cli/CLI/MainService.Logger.cs (hunk 10 above)
Comment (shargon) [reply neo-cli/CLI/MainService.Logger.cs:115]: Ping @Liaojinghui

**Comment by superboyiii** [comment]
@shargon Could you review again? Tested OK.

**Comment by Liaojinghui** [comment]
@shargon Hey shargon, I am not sure what exactly is your concern, just answer the following two previously unaddressed questions:

> Only one `Console.Write`?

Updated into one line.

> Now that it's all the same color, do we still need to parse the message?

Indeed we still need to parse the message, it is a historic issue as the previous log has messages sticked together while we need to separate them apart.

**Review by shargon — APPROVED** [review]

**Review by superboyiii — APPROVED** [review]

**Review by vncoelho — APPROVED** [review]

**Comment by vncoelho** [comment]
> I like `console log on` and `console log off`. But I still prefer the original whole line in the same color, which looks more intuitive.

@erikzhang, your review is blocking.
Can we proceed like this or you still want it to be off as standard?

**Comment by vncoelho** [comment]
@Liaojinghui, I think there should be some command to remove colors and keep original compatibility.
Perhaps some people will prefer it clean as the original one.

**Comment by Liaojinghui** [comment]
> @Liaojinghui, I think there should be some command to remove colors and keep original compatibility. Perhaps some people will prefer it clean as the original one.

It is already in the black/white mode for the most log, only errors will show an red [Error] and warning a yellow [warning], the rest remains black/white

**Comment by vncoelho** [comment]
Let me try again. Last time I saw it differently.

File: neo-cli/CLI/MainService.Logger.cs [hunk 15]
```diff
@@ -71,43 +92,80 @@ private void OnLog(string source, LogLevel level, object message)
...
-                        case LogLevel.Warning: WarningColor.Apply(); break;
+                        case LogLevel.Debug: logColor = DebugColor; logLevel = "DEBUG"; break;
+                        case LogLevel.Error: logColor = ErrorColor; logLevel = "ERROR"; break;
+                        case LogLevel.Fatal: logColor = FatalColor; logLevel = "FATAL"; break;
+                        case LogLevel.Info: logColor = InfoColor; logLevel = "INFO"; break;
+                        case LogLevel.Warning: logColor = WarningColor; logLevel = "WARN"; break;
+                        default: logColor = InfoColor; logLevel = "INFO"; break;
+                    }
+                    logColor.Apply();
+                    Console.Write($"{logLevel} {log} \t{messages[0],-20}");
```
Comment (vncoelho) [inline neo-cli/CLI/MainService.Logger.cs:112]: What are these ones here and this if that checks Length > 20?

Maybe add some comment.

Comment (Liaojinghui) [reply neo-cli/CLI/MainService.Logger.cs:112]: If they are longer than 20, the longer one will not align that well., the rest message will be pushed backwards. But in most (actually i would say all) existing situations, it aligns pretty good.

**Comment by vncoelho** [comment]
Tested again with long lines and works fine.
Colors are just for warning and error and it is almost the same as default.
Easier for futures control and modifications as well.
